File: mobile-app/ios/SonicWalkscape/SonicWalkscape/Services/AnalyticsService.swift
//
//  AnalyticsService.swift
//  SonicWalkscape
//
//  Anonymous analytics tracking with GDPR compliance
//

import Foundation
import UIKit

// MARK: - Event Types

enum AnalyticsEventType: String {
    case tourStarted = "tour_started"
    case pointTriggered = "point_triggered"
    case tourCompleted = "tour_completed"
    case tourAbandoned = "tour_abandoned"
    case followUsClicked = "follow_us_clicked"
    case contactClicked = "contact_clicked"
    case donationLinkClicked = "donation_link_clicked"
}

// MARK: - Trigger Type

enum TriggerType: String {
    case gps = "gps"
    case manual = "manual"
}

// MARK: - Analytics Event

struct AnalyticsEvent: Codable {
    let name: String
    let anonymousId: String
    let tourId: String?
    let pointId: String?
    let language: String?
    let device: String
    let osVersion: String
    let timestamp: String
    let properties: [String: AnyCodableValue]?
}

// MARK: - AnyCodableValue for flexible properties

enum AnyCodableValue: Codable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let v = try? container.decode(String.self) { self = .string(v) }
        else if let v = try? container.decode(Int.self) { self = .int(v) }
        else if let v = try? container.decode(Double.self) { self = .double(v) }
        else if let v = try? container.decode(Bool.self) { self = .bool(v) }
        else { self = .string("") }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let v): try container.encode(v)
        case .int(let v): try container.encode(v)
        case .double(let v): try container.encode(v)
        case .bool(let v): try container.encode(v)
        }
    }
}

// MARK: - Analytics Service

class AnalyticsService {
    static let shared = AnalyticsService()

    private let anonymousIdKey = "analytics_anonymous_id"
    private let pendingEventsKey = "analytics_pending_events"
    private var pendingEvents: [AnalyticsEvent] = []
    private let batchSize = 10
    private let flushInterval: TimeInterval = 30

    private var flushTimer: Timer?

    private init() {
        loadPendingEvents()
        startFlushTimer()
    }

    // MARK: - Anonymous ID

    var anonymousId: String {
        if let id = UserDefaults.standard.string(forKey: anonymousIdKey) {
            return id
        }
        let newId = UUID().uuidString
        UserDefaults.standard.set(newId, forKey: anonymousIdKey)
        return newId
    }

    // MARK: - Device Info

    private var deviceInfo: String {
        "ios-\(UIDevice.current.model)"
    }

    private var osVersionInfo: String {
        "iOS \(UIDevice.current.systemVersion)"
    }

    // MARK: - Consent Check

    private var hasConsent: Bool {
        UserPreferencesManager.shared.analyticsEnabled
    }

    // MARK: - Track Events

    func trackTourStarted(tourId: String, language: String, triggerType: TriggerType) {
        track(
            event: .tourStarted,
            tourId: tourId,
            properties: [
                "triggerType": .string(triggerType.rawValue),
                "language": .string(language),
                "platform": .string("ios")
            ]
        )
    }

    func trackPointTriggered(tourId: String, pointId: String, triggerType: TriggerType) {
        track(
            event: .pointTriggered,
            tourId: tourId,
            pointId: pointId,
            properties: [
                "triggerType": .string(triggerType.rawValue)
            ]
        )
    }

    func trackTourCompleted(tourId: String, durationMinutes: Int, pointsVisited: Int, triggerType: TriggerType) {
        track(
            event: .tourCompleted,
            tourId: tourId,
            properties: [
                "durationMinutes": .int(durationMinutes),
                "pointsVisited": .int(pointsVisited),
                "triggerType": .string(triggerType.rawValue),
                "platform": .string("ios")
            ]
        )
    }

    func trackTourAbandoned(tourId: String, durationMinutes: Int, lastPointIndex: Int) {
        track(
            event: .tourAbandoned,
            tourId: tourId,
            properties: [
                "durationMinutes": .int(durationMinutes),
                "lastPointIndex": .int(lastPointIndex),
                "platform": .string("ios")
            ]
        )
    }

    func trackFollowUsClicked(tourId: String) {
        track(
            event: .followUsClicked,
            tourId: tourId,
            properties: nil
        )
    }

    func trackContactClicked(tourId: String, channel: String) {
        track(
            event: .contactClicked,
            tourId: tourId,
            properties: [
                "channel": .string(channel)
            ]
        )
    }

    func trackDonationLinkClicked(tourId: String) {
        track(
            event: .donationLinkClicked,
            tourId: tourId,
            properties: nil
        )
    }

    // MARK: - Core Track Method

    private func track(
        event: AnalyticsEventType,
        tourId: String? = nil,
        pointId: String? = nil,
        properties: [String: AnyCodableValue]?
    ) {
        guard hasConsent else {
            DebugLogger.log("Analytics disabled by user preference")
            return
        }

        let formatter = ISO8601DateFormatter()
        let analyticsEvent = AnalyticsEvent(
            name: event.rawValue,
            anonymousId: anonymousId,
            tourId: tourId,
            pointId: pointId,
            language: UserPreferencesManager.shared.preferredLanguage,
            device: deviceInfo,
            osVersion: osVersionInfo,
            timestamp: formatter.string(from: Date()),
            properties: properties
        )

        pendingEvents.append(analyticsEvent)
        savePendingEvents()

        DebugLogger.log("Analytics: Queued \(event.rawValue)")

        // Flush if batch size reached
        if pendingEvents.count >= batchSize {
            flush()
        }
    }

    // MARK: - Flush Events

    func flush() {
        guard !pendingEvents.isEmpty else { return }
        guard hasConsent else {
            pendingEvents.removeAll()
            savePendingEvents()
            return
        }

        let eventsToSend = pendingEvents
        pendingEvents.removeAll()
        savePendingEvents()

        Task {
            await sendEvents(eventsToSend)
        }
    }

    private func sendEvents(_ events: [AnalyticsEvent]) async {
        let urlString = "\(Constants.API.fullURL)/analytics/events"
        guard let url = URL(string: urlString) else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let payload = ["events": events]

        do {
            let encoder = JSONEncoder()
            request.httpBody = try encoder.encode(payload)

            let (_, response) = try await URLSession.shared.data(for: request)

            if let httpResponse = response as? HTTPURLResponse,
               (200...299).contains(httpResponse.statusCode) {
                DebugLogger.success("Analytics: Sent \(events.count) events")
            } else {
                // Re-queue events on failure
                await MainActor.run {
                    self.pendingEvents.append(contentsOf: events)
                    self.savePendingEvents()
                }
                DebugLogger.warning("Analytics: Failed to send, re-queued \(events.count) events")
            }
        } catch {
            // Re-queue events on error (offline support)
            await MainActor.run {
                self.pendingEvents.append(contentsOf: events)
                self.savePendingEvents()
            }
            DebugLogger.warning("Analytics: Network error, re-queued \(events.count) events")
        }
    }

    // MARK: - Persistence

    private func savePendingEvents() {
        if let data = try? JSONEncoder().encode(pendingEvents) {
            UserDefaults.standard.set(data, forKey: pendingEventsKey)
        }
    }

    private func loadPendingEvents() {
        if let data = UserDefaults.standard.data(forKey: pendingEventsKey),
           let events = try? JSONDecoder().decode([AnalyticsEvent].self, from: data) {
            pendingEvents = events
        }
    }

    // MARK: - Timer

    private func startFlushTimer() {
        flushTimer = Timer.scheduledTimer(withTimeInterval: flushInterval, repeats: true) { [weak self] _ in
            self?.flush()
        }
    }
}

File: android-app/app 2/src/main/java/com/bandite/sonicwalkscape/services/AnalyticsService.kt
package com.bandite.sonicwalkscape.services

import android.content.Context
import android.os.Build
import com.bandite.sonicwalkscape.data.api.AnalyticsEvent
import com.bandite.sonicwalkscape.data.api.AnalyticsEventsRequest
import com.bandite.sonicwalkscape.data.api.ApiClient
import com.bandite.sonicwalkscape.utils.Constants
import com.bandite.sonicwalkscape.utils.DebugLogger
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AnalyticsService @Inject constructor(
    @ApplicationContext private val context: Context,
    private val apiClient: ApiClient,
    private val preferencesManager: UserPreferencesManager
) {
    private val scope = CoroutineScope(Dispatchers.IO)
    private val pendingEvents = mutableListOf<AnalyticsEvent>()
    private var flushJob: Job? = null
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)

    // Anonymous ID (persisted)
    private val anonymousId: String by lazy {
        val prefs = context.getSharedPreferences("analytics", Context.MODE_PRIVATE)
        prefs.getString(Constants.Analytics.ANONYMOUS_ID_KEY, null) ?: run {
            val newId = UUID.randomUUID().toString()
            prefs.edit().putString(Constants.Analytics.ANONYMOUS_ID_KEY, newId).apply()
            newId
        }
    }

    // Event types
    object Events {
        const val TOUR_STARTED = "tour_started"
        const val POINT_TRIGGERED = "point_triggered"
        const val TOUR_COMPLETED = "tour_completed"
        const val TOUR_ABANDONED = "tour_abandoned"
        const val FOLLOW_US_CLICKED = "follow_us_clicked"
        const val CONTACT_CLICKED = "contact_clicked"
        const val DONATION_LINK_CLICKED = "donation_link_clicked"
    }

    init {
        startAutoFlush()
    }

    // ========== Event Tracking ==========

    fun trackTourStarted(tourId: String, language: String, triggerType: String) {
        track(
            Events.TOUR_STARTED,
            mapOf(
                "tourId" to tourId,
                "language" to language,
                "triggerType" to triggerType
            )
        )
    }

    fun trackPointTriggered(tourId: String, pointId: String, triggerType: String) {
        track(
            Events.POINT_TRIGGERED,
            mapOf(
                "tourId" to tourId,
                "pointId" to pointId,
                "triggerType" to triggerType
            )
        )
    }

    fun trackTourCompleted(tourId: String, durationMinutes: Int, triggerType: String) {
        track(
            Events.TOUR_COMPLETED,
            mapOf(
                "tourId" to tourId,
                "durationMinutes" to durationMinutes,
                "triggerType" to triggerType
            )
        )
    }

    fun trackTourAbandoned(tourId: String, durationMinutes: Int, lastPointIndex: Int) {
        track(
            Events.TOUR_ABANDONED,
            mapOf(
                "tourId" to tourId,
                "durationMinutes" to durationMinutes,
                "lastPointIndex" to lastPointIndex
            )
        )
    }

    fun trackFollowUsClicked(tourId: String) {
        track(Events.FOLLOW_US_CLICKED, mapOf("tourId" to tourId))
    }

    fun trackContactClicked(tourId: String, channel: String) {
        track(
            Events.CONTACT_CLICKED,
            mapOf(
                "tourId" to tourId,
                "channel" to channel
            )
        )
    }

    fun trackDonationLinkClicked(tourId: String) {
        track(Events.DONATION_LINK_CLICKED, mapOf("tourId" to tourId))
    }

    // ========== Core Tracking ==========

    private fun track(eventName: String, properties: Map<String, Any?>) {
        scope.launch {
            // Check consent
            if (!preferencesManager.analyticsEnabled.first()) {
                DebugLogger.info("Analytics disabled - dropping event: $eventName")
                return@launch
            }

            val event = AnalyticsEvent(
                eventName = eventName,
                anonymousId = anonymousId,
                timestamp = dateFormat.format(Date()),
                properties = properties + getDefaultProperties()
            )

            synchronized(pendingEvents) {
                pendingEvents.add(event)
                DebugLogger.info("Queued analytics event: $eventName")

                // Flush if batch is full
                if (pendingEvents.size >= Constants.Analytics.MAX_BATCH_SIZE) {
                    flushEvents()
                }
            }
        }
    }

    private fun getDefaultProperties(): Map<String, Any> {
        return mapOf(
            "platform" to "android",
            "device" to Build.MODEL,
            "osVersion" to "Android ${Build.VERSION.RELEASE}",
            "appVersion" to getAppVersion()
        )
    }

    private fun getAppVersion(): String {
        return try {
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            packageInfo.versionName ?: "unknown"
        } catch (e: Exception) {
            "unknown"
        }
    }

    // ========== Batch Submission ==========

    private fun startAutoFlush() {
        flushJob?.cancel()
        flushJob = scope.launch {
            while (isActive) {
                delay(Constants.Analytics.FLUSH_INTERVAL_SECONDS * 1000)
                flushEvents()
            }
        }
    }

    fun flushEvents() {
        scope.launch {
            val eventsToSend: List<AnalyticsEvent>
            synchronized(pendingEvents) {
                if (pendingEvents.isEmpty()) return@launch
                eventsToSend = pendingEvents.toList()
                pendingEvents.clear()
            }

            try {
                val response = apiClient.apiService.submitAnalyticsEvents(
                    AnalyticsEventsRequest(eventsToSend)
                )

                if (response.isSuccessful) {
                    DebugLogger.info("Submitted ${eventsToSend.size} analytics events")
                } else {
                    // Re-queue events on failure
                    synchronized(pendingEvents) {
                        pendingEvents.addAll(0, eventsToSend)
                    }
                    DebugLogger.warning("Failed to submit analytics: ${response.code()}")
                }
            } catch (e: Exception) {
                // Re-queue events on error
                synchronized(pendingEvents) {
                    pendingEvents.addAll(0, eventsToSend)
                }
                DebugLogger.error("Analytics submission error", e)
            }
        }
    }

    fun clearPendingEvents() {
        synchronized(pendingEvents) {
            pendingEvents.clear()
        }
    }
}
